### DISCUSSION TITLE
Threads

### DISCUSSION THREAD

**Discussion by donarturo11** [muse-sequencer/muse#1275] [Team Posts]
I have three questions:
1. Can all other classes use thread class from MusECore namespace instead pthread directly?
2. Can std::thread be used as temporary class for handle threads instead phtreads inside of thread class?
3. Why QThread is more preferable instead std::thread?

**Comment by terminator356** [comment]
Hello! We use pthreads in some places because those threads need to have real time priority.
We cannot migrate those threads to QThread-based threads because QThread cannot obtain real time priority
 as a normal user, according to my findings. I really do wish that it would, but it will not.

However there is some recent progress in MusE. From the ChangeLog 25.05.23:
>       Disk prefetch thread is NO LONGER REALTIME! It is now a default NORMAL thread. (My fault long ago).
>       Rewrote MusE::seqStart() so prefetch is started before audio. Also, wait until prefetch is running!
>       TESTED OK playing and recording complex songs. The prefetch cache maintained a high fill level.
>       TODO: Now that disk prefetch thread is no longer realtime, replace its pthreads with QThread,
>         for easier cross-compatibility!

The 'disk prefetch thread' referred to above is the class 'AudioPrefetch : public Thread', found in audioprefetch.h
It would take some careful work to migrate the class to QThread-based.

Another real time thread in MusE is the ALSA midi sequencer thread class 'MidiSeq : public Thread', found in midiseq.h
Note that on Windows, MusE's ALSA support would be disabled in the CMake configuration,
 and therefore this ALSA sequencer thread would be irrelevant. All of MusE's midi would occur through Jack Midi.

So, that's two threads that maybe could be ignored or eliminated, at least under Windows.
But beyond those two threads, there is still some other pthread-based code in MusE that would have to be migrated to QThread.
Somehow...

**Reply by donarturo11** [reply]
I read some about threads both in STL and QT. I found that STL has not priority but QThread has priority in methods (compare: [std::thread](https://en.cppreference.com/w/cpp/thread/thread) and [QThread](https://doc.qt.io/qt-5/qthread.html)). I looked too into [LMMS includes](https://github.com/LMMS/lmms/tree/master/include), where I noticed that classes, like AudioAlsa, have two parents: QThread and AudioDevice.

**Comment by spamatica** [comment]
I assume the question is for portability to Windows, where the QThread priority settings align with the operating system (i think).

I remember wondering a long time a go if we could change to QThreads, #ifdef LINUX and aquire the pthread /from/ the QThread (as it certainly is a pthread internally) and do the SCHED_FIFO settings only in this case?

I don't know a lot about std::thread, I assume it is also portable, don't know if it allows for SCHED_FIFO realtime settings.
